FEA add temperature scaling to `CalibratedClassifierCV`

virchan opened the pull request:
#### Reference Issues/PRs

Closes #28574

#### What does this implement/fix? Explain your changes.

This PR adds **temperature scaling** to scikit-learn's [<code>CalibratedClassifierCV</code>](https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html#sklearn.calibration.CalibratedClassifierCV):

Temperature scaling can be enabled by setting `method = "temperature"` in `CalibratedClassifierCV`:

```python
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.calibration import CalibratedClassifierCV
from sklearn.svm import LinearSVC

X, y = make_classification(random_state=42)

X_train, X_calib, y_train, y_calib = train_test_split(X, y, random_state=42)

clf = LinearSVC(random_state=42)
clf.fit(X_train, y_train)
cal_clf = CalibratedClassifierCV(clf, method="temperature").fit(X_train, y_train)
```

This method supports both binary and multi-class classification.


#### Any other comments?
Cc @adrinjalali, @lorentzenchr in advance.

virchan committed:
FEA add temperature scaling to `CalibratedClassifierCV`

virchan committed:
added changelog

virchan commented on sklearn/calibration.py:
```diff
@@ -887,6 +908,149 @@ def loss_grad(AB):
+    labels = labels.astype(int)
+    max_logits = np.reshape(np.max(logits, axis=1), (-1, 1))
+
+    def negative_log_likelihood(beta, logits, labels, max_logits):
```
The temperature scaling method minimises the **negative log-likelihood (NLL)** with respect to the inverse temperature $\displaystyle \beta = \frac{1}{T}$:

<img width="284" alt="nll" src="https://github.com/user-attachments/assets/8aaa49e4-2e97-484b-baff-20730009db1a" />

where

- $Z$ is the output of `decision_function` or `predict_proba`, with entries $z_{i,j}$.
- $z_{\langle i \rangle}$ is the logit of the correct class for the $i$-th sample.
- $w_i$ is the sample weight.

The derivative with respect to $\beta$ is given by:

<img width="263" alt="nll_derivative" src="https://github.com/user-attachments/assets/2b0e167b-afb5-45a5-81ea-e06a23da0714" />

where $\displaystyle m_i = \max_{j} z_{i,j}$ is substracted to ensure numerical stability.

virchan commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +413,44 @@ def test_sigmoid_calibration():
         _SigmoidCalibration().fit(np.vstack((exF, exF)), exY)
 
 
+def test_temperature_scaling(data):
```
This test verifies that temperature scaling **does not affect accuracy** and that **the optimised temperature is always positive**.

I also noticed that the Brier score may improve or worsen depending on the dataset and the classifier being calibrated. Therefore, I did not include temperature scaling in the [<code>test_calibration</code>](https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/tests/test_calibration.py#L66) function.

This seems align with the remark made on page 3245 of [Classifier calibration: a survey on how to assess and improve predicted class probabilities](https://link.springer.com/article/10.1007/s10994-023-06336-7).

If there are any test cases I should add, please let me know—I’d be happy to include them.

ogrisel replied:
I think (single-parameter) temperature scaling provably improves Brier score (or log-loss) by only reducing its calibration error term and not changing the refinement error term on datasets/tasks with very specific structures, e.g balanced binary classification problem where means of the classes are symmetric around zero and the covariances are identical as explained in section 5 of https://arxiv.org/html/2501.19195v1#S5.

We could write a test based on this analysis.

virchan reviewed (commented):
A follow-up to [my comment on the Array API](https://github.com/scikit-learn/scikit-learn/issues/28574#issuecomment-2666814933): I don't think we can support the Array API here, as [<code>scipy.optimize.minimize</code>](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) does not appear to support it.

If I missed anything, please let me know—I'd be happy to investigate further.

virchan committed:
updated docstring.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -887,6 +909,149 @@ def loss_grad(AB):
+    Parameters
+    ----------
+    predictions : ndarray of shape (n_samples,) or (n_samples, n_classes)
+        The decision function or predict proba for the samples.
```
Let's be explicit if we expect one or the other and change the code accordingly.

Here have the impression that this code always expects logits (that is the output of `decision_function` and never the output of `predict_proba`). If the latter, we would need to take the log of it (+ and epsilon to avoid nans).

This info is originally present in the `method_name` variable of:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/calibration.py#L440-L443

I think `_fit_calibrator` should propagate this info to the calibrator to avoid any ambiguity. 

Maybe we should also undo pre-processing such as:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/calibration.py#L455-L463

and let the calibrator do what they find most suitable with either kind of input.

Maybe the calibrators could expose explicit tags to make it explicit on whether they can handle multiclass natively, or if the caller should be in charge of reducing a multiclass calibration problem into binary calibration sub problems with the OvR hack we currently implement for isotonic calibration and Platt scaling.

EDIT: the existing `calibrator.__sklearn_tags___().input_tags` might be enough to express this:

```python

tags = calibrator.__sklearn_tags()
supports_multiclass_logits = tags.input_tags.two_d_array and not tags.input_tags.one_d_array
```

as we have the following for `IsotonicRegression`:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/isotonic.py#L513-L517

The `_SigmoidCalibrator` can be udpated accordingly.

ogrisel replied:
Alternatively, we can rename the ambiguous "predictions" variable every-where to "logits" and take the `log` of `predict_proba` + eps as early as possible.

But this would introduce a behavior change.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -887,6 +909,149 @@ def loss_grad(AB):
+        negative_log_likelihood,
+        np.array([beta_0]),
+        args=(logits, labels, max_logits),
+        method="L-BFGS-B",
```
But this would deserve some benchmarking to confirm.

virchan replied:
I ended up choosing `minimize` over `minimize_scalar`, even though it's more expensive (e.g., it requires computing the gradient).

This is because `minimize_scalar`  doesn’t allow us to provide an initial guess---`beta_0 = 1.0`---when optimising the inverse temperature `beta` using `method = "Bounded"`. Even for `method = "Brent"`, it expects the initial guess `xb` to be a local minimum between two other points: [<code>func(xb) < func(xa) and  func(xb) < func(xc)</code>](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize_scalar.html#scipy.optimize.minimize_scalar), which is impossible to determine beforehand when fitting the calibrator.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -942,6 +1107,67 @@ def predict(self, T):
+        The initial temperature for optimization.
+
+    T_ : float
+        The optimized temperature.
```
Maybe we can just expose inverse temperature (aka the beta values), both initial and fitted directly instead of keeping them private attributes. The I would drop the `T_initial` parameter and `T_` attribute.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -942,6 +1107,67 @@ def predict(self, T):
+
+    def __init__(self, T_initial=1.0):
+        self.T_initial_ = T_initial
+        self._beta_0 = np.divide(1.0, self.T_initial_)
```
The constructor of scikit-learn estimators should always store the input parameters unchanged with the same names:

https://scikit-learn.org/stable/developers/develop.html#instantiation

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +413,44 @@ def test_sigmoid_calibration():
+    clfs = [
+        make_pipeline(StandardScaler(), LinearSVC(random_state=42)),
+        LogisticRegression(),
+    ]
```
Let's pass this as an argument of the test using `@pytest.mark.parametrize` instead.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -712,10 +728,11 @@ class _CalibratedClassifier:
     classes : array-like of shape (n_classes,)
```
The docstring of the `calibrators` parameter needs to be updated since the number of calibrators depend differently on whether the calibrator class can support multiclass or binary logits as inputs.

ogrisel commented on sklearn/calibration.py:
```diff
@@ -887,6 +909,149 @@ def loss_grad(AB):
+
+
 class _SigmoidCalibration(RegressorMixin, BaseEstimator):
     """Sigmoid regression model.
```
We should override `__sklearn_tags__` to set the same input tags as `IsotonicRegression`:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/isotonic.py#L513-L517

See:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/utils/_tags.py#L20-L25

ogrisel reviewed (commented):
Thanks for the PR. Here is a first pass of feedback:

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Updated docstrings of `CalibratedClassifierCV` and `_CalibratedClassifier`.
Updated constructor of `_TemperatureScaling` class.
Updated `test_temperature_scaling` in `test_calibration.py`.
Added `__sklearn_tags__` to `_TemperatureScaling` class.

virchan committed:
Fix typo in `_TemperatureScaling`'s `fit` method.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -683,12 +711,19 @@ def _fit_calibrator(clf, predictions, y, classes, method, sample_weight=None):
-            calibrator = _SigmoidCalibration()
-        calibrator.fit(this_pred, Y[:, class_idx], sample_weight)
+
+    if (method == "isotonic") or (method == "sigmoid"):
```
```suggestion
    if method in ("isotonic", "sigmoid"):
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -683,12 +711,19 @@ def _fit_calibrator(clf, predictions, y, classes, method, sample_weight=None):
+    elif method == "temperature":
+        calibrator = _TemperatureScaling()
+        calibrator.fit(predictions, Y, sample_weight)
         calibrators.append(calibrator)
```
```suggestion
        calibrators.append(calibrator)
    else:
        raise ValueError("Parameter method must be in ...[fill me]")
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +961,150 @@ def loss_grad(AB):
+    labels = labels.astype(int)
+    max_logits = np.reshape(np.max(logits, axis=1), (-1, 1))
+
+    def negative_log_likelihood(beta, logits, labels, max_logits):
```
We should reuse what we already have, e.g. in `sklearn._loss`.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Updated test cases for temperature scaling in `test_calibration.py`. Updated doc-strings of temperature scaling in `calibration.py`. Updated formatting.

virchan commented on sklearn/calibration.py:
```diff
@@ -887,6 +970,137 @@ def loss_grad(AB):
+        return np.astype(np.hstack([-decision_values, decision_values]), np.float64)
+
+
+def _temperature_scaling(predictions, labels, sample_weight=None, beta_0=1.0):
+    """Probability Calibration with temperature scaling
+
+    Parameters
+    ----------
+    predictions : ndarray of shape (n_samples,) or (n_samples, n_classes)
+        The output of `decision_function` or `predict_proba`. If the input
+        appears to be probabilities (i.e., values between 0 and 1 that sum to 1
+        across classes), it will be converted to logits using `np.log(p + eps)`.
+
+        Binary decision function outputs (1D) will be converted to two-class
+        logits of the form (-x, x). For shape (n_samples, 1), the same applies.
```
> Let's be explicit if we expect one or the other and update the code (e.g. variable names) accordingly.
> 
> Here have the impression that this code always expects logits (that is the output of `decision_function` and never the output of `predict_proba`). If the latter, we would need to take the log of it (+ and epsilon to avoid nans).
> 
> [...]

Let me know if this works---happy to keep refining it if needed.

virchan commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +418,74 @@ def test_sigmoid_calibration():
+    # The optimized temperature should always be positive
+    assert cal_clf.calibrated_classifiers_[0].calibrators[0].beta > 0
+
+    # Refinement error should be invariant under temperature scaling.
+    # Use ROC AUC as a proxy for refinement error.
+    roc_auc_diff = np.abs(
+        roc_auc_score(y_test, y_scores_cal, multi_class="ovr")
+        - roc_auc_score(y_test, y_scores, multi_class="ovr")
+    )
+
+    assert roc_auc_diff <= 0.03
```
> I think (single-parameter) temperature scaling provably improves Brier score (or log-loss) by only reducing its calibration error term and not changing the refinement error term on datasets/tasks with very specific structures, e.g balanced binary classification problem where means of the classes are symmetric around zero and the covariances are identical as explained in section 5 of https://arxiv.org/html/2501.19195v1#S5.
> 
> We could write a test based on this analysis.

I ended up adding this test to check that the refinement error remains unchanged after temperature scaling.

I used the area under the curve as a proxy for the refinement error, based on this Stack Exchange post:

https://stats.stackexchange.com/questions/631440/what-is-the-relationship-between-the-brier-score-refinement-and-the-area-under

Other than that, I'm not sure how to compute or approximate the refinement error using scikit-learn's functions.

virchan reviewed (commented):
I'm still working on addressing the feedback, but I also wanted to share some findings related to it and provide an update.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Fix failing `test_float32_predict_proba`.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +977,142 @@ def loss_grad(AB):
+    if sample_weight is not None:
+        sample_weight = np.astype(sample_weight, logits.dtype)
+
+    loss = HalfMultinomialLoss()
```
```suggestion
    loss = HalfMultinomialLoss(n_classes=XXX)
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +977,142 @@ def loss_grad(AB):
+            The negative log likelihood loss.
+
+        """
+        # Select the logit of the correct class
+        raw_prediction = np.astype(softmax(beta * logits), predictions.dtype)
```
This seems wrong. It should be as simple as `raw_prediction = beta * logits`. And the logits should already have a suitable dtype.

virchan replied:
You are right. I missed that because I didn’t read the docstring carefully enough:

```console
loss_i = log(sum(exp(raw_pred_{i, k}), k=0..n_classes-1))
             - sum(y_true_{i, k} * raw_pred_{i, k}, k=0..n_classes-1)
```

Thanks for spotting it!

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +977,142 @@ def loss_grad(AB):
+            y_true=y_true, raw_prediction=raw_prediction, sample_weight=sample_weight
+        )
+
+        g_true = g[np.arange(len(labels)), labels]
+        raw_prediction_true = raw_prediction[np.arange(len(labels)), labels]
+        gradient = (
+            g_true * raw_prediction_true * (1 - raw_prediction_true) * logits_true
+        )
```
```suggestion
       gradient = beta * g
```

lorentzenchr reviewed (commented):
I few computational things seem off.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update `HalfMultinomialLoss` docstring.
Update `minimize` in `_temperture_scaling` to `minimize.scalar`.
Update `test_calibration.py` to check the optimised inverse temperature is between 0.1 and 10.

virchan commented on sklearn/calibration.py:
```diff
@@ -887,6 +980,138 @@ def loss_grad(AB):
+
+        return l.sum()
+
+    beta_minimizer = minimize_scalar(beta_loss, bounds=(0.1, 10.0), method="bounded")
```
> Why not pass `bracket=(.5, 2)` or `bracket=(0.1, 10.)` since we work on a multiplicative scale around inverse temperature `beta_0=1.0`?
> 

I think this might work — I've set `bounds=(0.1, 10.0)` and used the `test_temperature_scaling` test function to confirm that the optimised beta stays within that range.

That said, I do wonder if users might ask for a way to set a custom initial guess `beta_0` when performing temperature scaling.

virchan commented on sklearn/calibration.py:
```diff
@@ -887,6 +980,138 @@ def loss_grad(AB):
+            The negative log likelihood loss.
+
+        """
+        # Ensure raw_prediction has the same dtype as labels using .astype().
+        # Without this, dtype promotion rules differ across NumPy versions:
+        #
+        #   beta = np.float64(0)
+        #   logits = np.array([1, 2], dtype=np.float32)
+        #
+        #   result = beta * logits
+        #   - NumPy < 2: result.dtype is float32
+        #   - NumPy 2+:  result.dtype is float64
+        #
+        #  This can cause dtype mismatch errors downstream (e.g., buffer dtype).
+        raw_prediction = xp.astype(beta * logits, dtype_)
```
> And the logits should already have a suitable dtype.

I had to explicitly set the dtype here—otherwise `scipy.optimize.minimize_scalar` fails when 
`(beta * logits).dtype == np.float32`. 

This issue is caught by our own [<code>test_float32_predict_proba</code>](https://github.com/scikit-learn/scikit-learn/blob/7131d9488dfb8edd6ae042caca57dd76523f395b/sklearn/tests/test_calibration.py#L1085C29-L1086C5).

I noticed Sigmoid Calibration [handles the same issue](https://github.com/scikit-learn/scikit-learn/blob/7131d9488dfb8edd6ae042caca57dd76523f395b/sklearn/calibration.py#L852-L855), so I reused their fix and comment for consistency.

Happy to adjust if I missed anything!

virchan reviewed (commented):
There are some CI failures—I'll fix those shortly.

Also considering adding a `verbose` parameter to `CalibratedClassifierCV` to optionally display convergence info when optimising the inverse temperature `beta`.

virchan committed:
Update `_TemperatureScaling` tags.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Add `test_calibration_method` in `test_calibration.py` to check invalid `method` parameter.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update `__sklearn_tags__` for the `_TemperatureScaling` class to the default tags.

virchan commented on sklearn/calibration.py:
```diff
@@ -942,6 +1167,81 @@ def predict(self, T):
+        X = _standardize_decision_values(X)
+        return softmax(self.beta * X)
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        return tags
```
> We should override `__sklearn_tags__` to set the same input tags as `IsotonicRegression` [...]

At this point, I'm using the default `tags` to suppress CI failures.

But it's probably safer to set the following explicitly, as originally suggested:

```python
tags.input_tags.one_d_array = True
tags.input_tags.two_d_array = True
```

(This is different from the `tags` used in `IsotonicRegression`.) 
However, I'm not sure what test(s) should be added to `test_calibration.py` to check the `tags`, so that CI doesn't complain…

The only case where the `_TemperatureScaling` class receives a 1D array is when the input comes from the `decision_function` in binary classification. For example, the output of:

```python
SVC(probability = False).decision_function(X_train)
```

I've defined the private function `_standardize_decision_values` to ensure the input array is converted to 2D when calling the `fit` and `predict` methods. So I think keeping the default `tags.input_tags.one_d_array = False` is still valid.

If I missed anything, I'd be happy to continue working on it and am open to any suggestions!

OmarManzoor commented on sklearn/calibration.py:
```diff
@@ -887,6 +980,138 @@ def loss_grad(AB):
+        #  This can cause dtype mismatch errors downstream (e.g., buffer dtype).
+        raw_prediction = xp.astype(beta * logits, dtype_)
+
+        l = halfmulti_loss.loss(y_true=labels, raw_prediction=raw_prediction)
```
Since the loss module is implemented in Cython will using the array api work here for example with PyTorch?

lorentzenchr replied:
No, it won't.

OmarManzoor replied:
Yes that makes sense. I think then in the initial inclusion of this feature we should avoid introducing the array api here.

OmarManzoor reviewed (commented):
Thanks for the PR @virchan 
I added a few comments

virchan committed:
Update docstrings by applying code-review suggestions.

virchan committed:
Update docstrings by applying code-review suggestions.

virchan committed:
Updated the `predict_proba` method of `_CalibratedClassifier` by applying suggestions from code-reviews.

virchan committed:
Update docstring of the `_temperature_scaling` function by applying suggestions from code reviews.

virchan commented on sklearn/calibration.py:
```diff
@@ -58,7 +63,7 @@
 
 class CalibratedClassifierCV(ClassifierMixin, MetaEstimatorMixin, BaseEstimator):
-    """Probability calibration with isotonic regression or logistic regression.
+    """Calibrate probabilities using isotonic, logistic, or temperature scaling.
```
I like your suggestion on using "sigmoid". Thank!

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update the `_standardize_decision_values` function to convert the output to `numpy.float64`. Remove Array API from the `_temperature_scaling` function.

OmarManzoor commented on sklearn/calibration.py:
```diff
@@ -942,6 +1151,83 @@ def predict(self, T):
+
+    Attributes
+    ----------
+    beta_0 : float
```
beta_0 feels more like a parameter.

virchan committed:
Replace `numpy.astype` with `.astype` in `_temperature_scaling` function. Update docstring of the `_TemperatureScaling` class by moving `beta_0` to the "Parameters" section.

virchan committed:
Merge branch 'main' into issues/28574_implementing_temperature_scaling

virchan commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,126 @@ def loss_grad(AB):
+        decision_values = decision_values.reshape(-1, 1)
+        logits = np.hstack([-decision_values, decision_values])
+
+    return logits.astype(np.float64)
```
Originally, I used the `_find_matching_floating_dtype` function from our Array API to ensure that the inputs were converted to the same `dtype` when fitting:

```python
_TemperatureScaling().fit(X, y, sample_weight)
```

This was to handle the `float32` and `int` test cases (as in the `test_float32_predict_proba` and `test_temperature_scaling_input_validation` tests).

Since we decided not to introduce the Array API here, I’ve updated the private helper function to convert the output to `np.float64`, so that `X`, `y`, and `sample_weight` are always `np.float64` when minimising `beta`.

Happy to adjust this further if needed!

OmarManzoor replied:
I think this makes sense when we consider the `int` dtype. But for float32 shouldn't we let our arrays remain that way if they were in float32 initially, considering that the HalfMultinomialLoss does support float32. Or do we want to keep the computations here in float64 for some consistency and don't really want to preserve the dtype?

virchan replied:
I agree that we should keep the arrays' dtype as they were initially, especially the case of `float32` and `float64`.

The only case where input's dtype is `int` that I've encountered is in the newly added `test_temperature_scaling_input_validation function`. I copied this from [<code>test_isotonic_2darray_more_than_1_feature</code>](https://github.com/scikit-learn/scikit-learn/blob/efe3b63589f628516a16ac9d620ccf172052572f/sklearn/tests/test_isotonic.py#L639C5-L639C46) in `test_isotonic.py` to mimic the case of calibrating the output of `decision_function` in binary classification. But multiplying by the inverse temperature `beta` will convert `int` to `float`, which causes `scipy.optimize.minimize_scalar` to raise an error. (Because the labels have `int` dtype.) That's why I made the helper function `_standardize_decision_values` always return `np.float64` in the previous version.

Thinking about this more, I believe it's safe to assume that the inputs to `_temperature scaling` always have a floating-point dtype, since they are meant to be outputs of `decision_function` or `predict_proba`. So, I've updated the test input in `test_temperature_scaling_input_validation` to use `np.float64`.

We can also check the inputs' dtype and raise an error if it is neither `np.float32` or `np.float64`.

Alternatively, I noticed that we don't test `_SigmoidCalibration` on 1D arrays (unlike isotonic). So maybe we could just remove the `test_temperature_scaling_input_validation` function...

virchan committed:
Update dtype convertion in the `_temperature_scaling` function. Update input dtype to `numpy.float64` in `test_tempereature_scaling_input_validation` function.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Merge remote-tracking branch 'origin/issues/28574_implementing_temperature_scaling' into issues/28574_implementing_temperature_scaling

OmarManzoor commented on sklearn/calibration.py:
```diff
@@ -942,6 +1162,85 @@ def predict(self, T):
+        The optimized inverse temperature.
+    """
+
+    def __init__(self, beta_0=1.0):
```
Is this really configurable? I don't see how we can change this initial value as this doesn't seem to be a parameter in CalibratedClassifierCV and the places from where `_TemperatureScaling` is called (including the tests) seem to only use the default?

virchan replied:
I removed the `beta_0` parameter as I don't see a good way to configure it without changing the API. I also think it's better not to mention `beta_0` at all, since temperature scaling is about modifying the standard softmax function.

OmarManzoor commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +443,74 @@ def test_sigmoid_calibration():
+    assert roc_auc_diff <= 0.02
+
+
+def test_temperature_scaling_input_validation():
+    # Check that _TemperatureScaling can handle 2d-array with only 1 feature
+    X = np.arange(10).astype(np.float64)
```
Maybe test this with both float64 and float32?
```suggestion
def test_temperature_scaling_input_validation(global_dtype):
    # Check that _TemperatureScaling can handle 2d-array with only 1 feature
    X = np.arange(10).astype(global_dtype)
```

OmarManzoor commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +443,74 @@ def test_sigmoid_calibration():
+        - roc_auc_score(y_test, y_scores, multi_class="ovr")
+    )
+
+    assert roc_auc_diff <= 0.02
```
Should we also add this:
`# Check that brier score has improved after calibration`

OmarManzoor reviewed (commented):
Thanks for the updates @virchan 
This mostly looks good now. I added a few more comments.
Also we might need to the update the user guide on calibration to include temperature scaling but maybe that can be done in a follow up PR.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -683,14 +712,28 @@ def _fit_calibrator(clf, predictions, y, classes, method, sample_weight=None):
+    else:
+        raise ValueError(
+            f"Invalid method '{method}'."
+            "Parameter `method` must be one of"
```
```suggestion
            "Parameter method must be one of"
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -755,26 +832,38 @@ def predict_proba(self, X):
+                proba[:, 0] = 1.0 - proba[:, 1]
+            else:
+                denominator = np.sum(proba, axis=1)[:, np.newaxis]
+                # In the edge case where for each class calibrator returns a null
```
```suggestion
                # In the edge case where for each class calibrator returns a zero
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -755,26 +832,38 @@ def predict_proba(self, X):
+                proba = np.divide(
+                    proba, denominator, out=uniform_proba, where=denominator != 0
+                )
+
```
```suggestion
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -755,26 +832,38 @@ def predict_proba(self, X):
+
+        elif self.method == "temperature":
+            proba = self.calibrators[0].predict(predictions)
+
```
```suggestion
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
 
 
+def _standardize_decision_values(decision_values, eps=1e-8):
+    """Convert decision_function values to 2D and predict_proba values to logits
```
```suggestion
    """Convert decision_function values to 2D and predict_proba values to logits.
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+    logits : ndarray of shape (n_samples, n_classes)
+    """
+    if (decision_values.ndim == 2) and (decision_values.shape[1] > 1):
+        # Check if it is the output of `predict_proba`
```
```suggestion
        # Check if it is the output of predict_proba
```
It's a comment, so no `` required.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
     return AB_[0] / scale_constant, AB_[1]
 
 
+def _standardize_decision_values(decision_values, eps=1e-8):
```
```suggestion
def _convert_to_logits(decision_values, eps=1e-8):
```
The word standardize is a bit odd and usually means a different thing.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+
+
+def _temperature_scaling(predictions, labels, sample_weight=None, beta_0=1.0):
+    """Probability Calibration with temperature scaling
```
```suggestion
    """Probability Calibration with temperature scaling.
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+        decision_values = decision_values.reshape(-1, 1)
+        logits = np.hstack([-decision_values, decision_values])
+
+    return logits.astype(dtype=decision_values.dtype)
```
```suggestion
    return logits
```
The `check_array` at the beginning simplifies this.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+    -------
+    logits : ndarray of shape (n_samples, n_classes)
+    """
+    if (decision_values.ndim == 2) and (decision_values.shape[1] > 1):
```
```suggestion
    decision_values = check_array(decision_values, dtype=[np.float64, np.float32])
    if (decision_values.ndim == 2) and (decision_values.shape[1] > 1):
```
Needs a `from .utils.validation import check_array`

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+    C. Guo, G. Pleiss, Y. Sun, & K. Q. Weinberger, ICML 2017.
+    """
+    check_consistent_length(predictions, labels)
+    logits = _standardize_decision_values(predictions)
```
```suggestion
    logits = _standardize_decision_values(predictions)  # guarantees np.float64 or np.float32
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+        sample_weight=sample_weight, n_classes=logits.shape[1]
+    )
+
+    def beta_loss(beta=beta_0):
```
```suggestion
    def log_loss(log_beta=log_beta_0):
```

OmarManzoor replied:
So we purposely treat the original beta as log_beta to change the formulation of the optimization problem?

lorentzenchr replied:
Now I get what you mean. I changed the suggestion accordingly.

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+
+        Parameters
+        ----------
+        beta : float
+            The current inverse temperature value during optimisation.
```
```suggestion
        log_beta : float
            The current logarithm of the inverse temperature value during optimisation.
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+        #   - NumPy 2+:  result.dtype is float64
+        #
+        #  This can cause dtype mismatch errors downstream (e.g., buffer dtype).
+        raw_prediction = (beta * logits).astype(dtype_)
```
```suggestion
        raw_prediction = (np.exp(log_beta) * logits).astype(dtype_)
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+    )
+
+    def beta_loss(beta=beta_0):
+        """Compute the negative log likelihood loss and its derivative
+            with respect to temperature.
```
```suggestion
        """Compute the log loss as a parameter of the inverse temperature (beta).
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+
+        return l.sum()
+
+    beta_minimizer = minimize_scalar(beta_loss, bounds=(0.1, 10.0), method="bounded")
```
```suggestion
    log_beta_minimizer = minimize_scalar(beta_loss, bounds=(-10.0, 10.0))
```

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -887,6 +976,137 @@ def loss_grad(AB):
+
+    beta_minimizer = minimize_scalar(beta_loss, bounds=(0.1, 10.0), method="bounded")
+
+    return beta_minimizer.x
```
```suggestion
    return np.exp(log_beta_minimizer.x)
```

virchan committed:
Update `calibration.py` and `test_calibration.py` by applying suggestions from code reviews.

lorentzenchr commented:
@virchan While a certain time delay is not an issue, it is a bit strange that all my review comments are marked as resolved but no new commits were pushed.

virchan commented:
Thank you for checking in @lorentzenchr!

I'm currently working on verifying that the log loss is being optimised. The next update should be within the next two days.

virchan committed:
Update `test_calibration.py` to check log loss reductions, AUC improvements, and accuracy invarance for temperature scaling when `ensemble = False`.

virchan committed:
Remove the `beta_0` parameter in temperature scaling.

virchan committed:
Add temperature scaling to user guide.

virchan committed:
Merge branch 'main' into issues/28574_implementing_temperature_scaling

virchan committed:
Removed DecisionTreeClassifier from the `test_temperature_scaling` function.

virchan reviewed (commented):
TL;DR: I updated the user guide and explored workarounds for checking the properties of temperature scaling.

I've learned how cross-validation affects the behaviour of temperature scaling in `CalibratedClassifierCV`.

When `ensemble=True`, fitting the model creates `cv` classifier–calibrator pairs:

`(estimator_1, TS_1), ... (estimator_cv, TS_cv)`

So we end up with `cv` separate temperature scaling calibrators, each with its own `beta`. This is well explained in the docstring of the `calibrated_classifiers_` attribute.

When calling `CalibratedClassifierCV.predict_proba()`, we're actually averaging the outputs of each `TS_i`, which are computed from the logits produced by the corresponding `estimator_i`, trained on the `i`-th fold.

That's why I wasn’t getting the expected test results: the averaged temperature-calibrated probabilities don’t necessarily minimise log loss, may affect accuracy, and can even worsen ROC-AUC scores.

My current workaround is to test the following:

1. Accuracy invariance
2. Log loss improvement (comparing with the standard softmax)
3. ROC-AUC score improvement

only when `ensemble=False`, because in that case there's a single (hence canonical) `beta`. Averaging probabilities makes sense across folds, but expecting the averaged probabilities to satisfy all of these properties at once feels overly optimistic.

An alternative is to refit and find the optimal `beta` using the full training set. However, optimising for log loss can change accuracy (and vice versa), so this still doesn't resolve the trade-off. As I think more about it, there doesn't seem to be a natural way to choose an optimal `beta` across CV folds that satisfies all of the above properties.

Another option is to only allow `method="temperature"` when `ensemble=False`. That’s the least effort solution, but it comes at the cost of losing the benefits of parallelism and cross-validation.

Overall, I think testing these properties under `ensemble=False` is a fair compromise.

Let me know if I've overlooked anything---I'll keep working on it!

lorentzenchr reviewed (commented):
Not far away from the finish line.

virchan committed:
Updated comments in the `log_loss` function inside the `_temperature_scaling` function in `calibration.py`. Updated comments in the `test_temperature_scaling` function in `test_calibration.py`.

virchan committed:
Added `_check_response_method` to `CalibratedClassifierCV.fit`, `_fit_classifier_calibrator_pair`, and `_CalibratedClassifier.predict_proba` to ensure `estimator.predict_proba` has shape (n_samples, 2) in binary case.

virchan committed:
Updated the `test_temperature_scaling` function to train the calibrator(s) on calibrating sets.

virchan committed:
Updated the `test_calibration_method_raises` function to test error raising for user-facing API instead of private functions.

virchan committed:
Updated tags for the `_TemperatureScaling` class.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Fixed typos in the `test_temperature_scaling` function.

virchan commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +446,85 @@ def test_sigmoid_calibration():
+        y_scores_softmax = softmax(_convert_to_logits(predictions))
+        y_scores_cal = cal_clf.predict_proba(X_test)
+        # Log Loss should be improved since it is the loss function
+        # in temperature scaling
```
I see what you're saying.

This time, I've mimicked the example from the `CalibratedClassifierCV` docstring to fit the calibrator on the calibrating set using a frozen estimator (fitted on the training set).

When `ensemble = False`, the improvements of log loss and ROC AUC behave as expected on the calibrating set.

Let me know what you think.

virchan reviewed (commented):
At this point, CI is complaining about missing test coverage for `tags` and for the errors raised in `_fit_calibrator` and `_CalibratedClassifier`.

I'm not quite sure how to handle the `tags` part, but I think the other two cases are fine, since the tests should cover warnings raised by the user-facing API, rather than by private functions.

I'll work on the convergence warning for `minimize_scalar` later.

virchan committed:
Fixed typos in the user guide `calibration.rst`

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update the `_temperature_scaling` function to display non-convergence message.

virchan committed:
Tighten absolute error in solution `xopt` acceptable for convergence in the `_temperature_scaling` function.

virchan reviewed (commented):
Updates on Raising Warnings When `scipy-optimize.minimize_scalar` Doesn't Converge.

`minimize_scalar(method = "bounded")` raises the following warnings when optimisation is unsuccessful:

1. "Maximum number of function calls reached". Controled by the `"maxiter"` in the `options` dictionary.
2. "Nan result encountered".

(Reference: [here](https://github.com/scipy/scipy/blob/f2d22b2308bc0269e18fe2f27fe6ab17540a4ac9/scipy/optimize/_optimize.py#L2429-L2434))

For (1), the default value is `maxiter=500`. In my testing (over about a week of trial and error), setting `maxiter=40` or lower may trigger non-convergence. However, reducing the number of iterations negatively affects the performance of the calibrator in otherwise converging cases. So I don't think it's worth lowering maxiter just to surface a warning in rare failure cases.

For (2), the docstring of `HalfMultinomialLoss` says:

```python
        loss_i = log(sum(exp(raw_pred_{i, k}), k=0..n_classes-1))
                - sum(y_true_{i, k} * raw_pred_{i, k}, k=0..n_classes-1)
```

As far as I can tell, the only two ways `loss_i` could become `NaN` are:

- if `raw_pred` contains `np.inf`, or is entirely `-np.inf`.
- if `raw_pred` contains `NaN`.

But in either case, scikit-learn's [<code>check_array</code>](https://github.com/scikit-learn/scikit-learn/blob/bb56546fa589950335951a372a0ae9f3638e8834/sklearn/utils/validation.py#L734) would catch this and raise an error before `minimize_scalar` even runs.

There's also the edge case where the minimiser lands exactly on a boundary ($\pm 10$), but due to the xtol tolerance in SciPy's optimizer, it ends up returning something like $\pm 9.9999999…$ and still counts as a successful termination.

So overall, I'm leaning towards our current implementation being fine in terms of convergence handling. That said, I might be missing some edge cases or subtle points.

Just to be safe, I've updated the `options` dictionary in `_temperature_scaling` to show a warning when convergence fails, and I've also tightened the absolute tolerance `xatol` for convergence.

Let me know if I've overlooked anything — I'll keep working on it!

Ball-Man commented:
If I can add a minor comment, while testing it I noticed that the learned temperature is treated as a plain attribute (`_TemperatureScaling.beta`). As a result, even after fitting, the `_TemperatureScaling` instance is detected as not fitted by `check_is_fitted`. This is easily solved by treating temperature as a learned parameter instead, with the trailing underscore syntax (e.g.  `_TemperatureScaling.beta_`). I guess this may be superfluous, since it's an internal estimator. Still, the sigmoid calibrator (as a counterexample) does behave appropriately, learning the parameters `a_` and `b_`. Thanks for your contribution.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Replace the `beta` attribute in temperature scaling with the `beta_` attribute to ensure `check_is_fitted` runs correctly.

virchan reviewed (commented):
Thank you for the review, @Ball-Man!

I can confirm that scikit-learn's [<code>check_is_fitted</code>](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html) function raises an error for temperature scaling, but not for sigmoid or isotonic calibration, after fitting on a calibration dataset.

I've updated the attribute as you suggested.

virchan committed:
Mark the `ValueError` lines in `_fit_calibrator` and `_CalibratedClassifier.predict_proba()` as `# pragma: no cover`.

virchan committed:
Raise RuntimeError when temperature scaling fails to optimize.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update `test_temperature_scaling` to calibrate SVC.

virchan committed:
Update `test_temperature_scaling` to fix CI failure.

virchan reviewed (commented):
I realised that `# pragma: no cover` can be used to raise errors without affecting CI coverage. So I've added it to both `_CalibratedClassifier.predict_proba()` and `_fit_calibrator`.

Specifically, if the `method` parameter is anything other than `"sigmoid"`, `"isotonic"`, or `"temperature"`, the input validation in `CalibratedClassifierCV` will raise an error before any of these private functions are even called.

Similarly, I was able to raise an error when `minimize_scalar` fails to converge, without upsetting CI coverage in the absence of a dedicated test case. So this fix felt reasonable.

Let me know your thoughts, I'm happy to keep working on this!

OmarManzoor reviewed (commented):
A few comments otherwise looks good. Thank you for the updates @virchan

OmarManzoor commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +425,83 @@ def test_sigmoid_calibration():
+    X_2d = X.reshape(-1, 1)
+    y = np.random.randint(0, 2, size=X.shape[0])
+
+    ts = _TemperatureScaling().fit(X, y)
+    ts_2d = _TemperatureScaling().fit(X_2d, y)
+
+    y_pred1 = ts.predict(X)
+    y_pred2 = ts_2d.predict(X_2d)
+
+    assert_allclose(y_pred1, y_pred2)
+
```
Maybe over here just call `get_tags(ts)` and assert the two tags that you have specified in this class. You would need to import the get_tags method. This will remove that code coverage warning.

virchan replied:
Calling `get_tags` does remove the code coverage warning. Thanks, @OmarManzoor!

Would you also like me to add this for `isotonic` and `sigmoid` (including `tags`)? I don't see this in their test cases but code coverage doesn't complain about it.

OmarManzoor replied:
I think we can leave those for this particular PR

virchan committed:
update `test_temperature_scaling` function to check accuracy invariance in all `ensemble` cases.

virchan committed:
update `CalibratedClassifierCV` docstrings by applying reviewer's suggestions.

virchan committed:
add `check_is_fitted` to the `test_temperature_scaling` function.

virchan committed:
add `get_tags` to the `test_temperature_scaling_input_validation` function.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
update the `test_temperature_scaling` function to fix CI failure.

virchan committed:
update `test_calibration.py` to trigger CI re-run.

virchan committed:
update `test_calibration.py` to trigger CI re-run.

virchan commented on sklearn/calibration.py:
```diff
@@ -683,41 +736,32 @@ def _fit_calibrator(clf, predictions, y, classes, method, sample_weight=None):
+        calibrator.fit(predictions, y, sample_weight)
         calibrators.append(calibrator)
 
+    else:  # pragma: no cover
```
This is suggested in [#31068 (comment)](https://github.com/scikit-learn/scikit-learn/pull/31068#discussion_r2021232995). I also believe it's safer to raise an error in this case.

virchan replied:
Maybe we could ask @lorentzenchr to weigh in here again, in light of [#31068 (comment)](https://github.com/scikit-learn/scikit-learn/pull/31068#discussion_r2106926742)---whether it's worthwhile to also include tests for private functions.

I'm happy to follow whichever approach the team prefers.

virchan reviewed (commented):
CI passed!

OmarManzoor reviewed (approved):
LGTM. Thank you @virchan

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -942,6 +1149,79 @@ def predict(self, T):
+        X_ : ndarray of shape (n_samples, n_classes)
+             The predicted data.
+        """
+        X = _convert_to_logits(X)
+        return softmax(self.beta_ * X)
```
```suggestion
        logits = _convert_to_logits(X)
        return softmax(self.beta_ * logits)
```

lorentzenchr commented on sklearn/tests/test_calibration.py:
```diff
@@ -273,6 +290,7 @@ def multiclass_brier(y_true, proba_pred, n_classes):
     cal_clf.fit(X_train, y_train)
     cal_clf_probs = cal_clf.predict_proba(X_test)
     calibrated_brier = multiclass_brier(y_test, cal_clf_probs, n_classes=n_classes)
+    print(f"{calibrated_brier = }", f"{1.1 * uncalibrated_brier = }")
```
```suggestion
```

lorentzenchr commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +427,88 @@ def test_sigmoid_calibration():
+        calibrator = calibrated_classifier.calibrators[0]
+        # Should not raise any error
+        check_is_fitted(calibrator)
+        # The optimal inverse temperature parameter should always
+        # be positive
```
```suggestion
        # The optimal inverse temperature parameter should always be positive
```

lorentzenchr commented on doc/modules/calibration.rst:
```diff
@@ -276,6 +276,54 @@ probabilities, the calibrated probabilities for each class
+function with a temperature parameter :math:`T`:
+
+.. math::
+       \mathrm{softmax}\left( \left[ \frac{z_1}{T},
+       \cdots, \frac{z_n}{T} \right] \right) \,,
+
+where :math:`\left[z_1, \cdots, z_n \right]` are the logits for each class produced
```
```suggestion
       \mathrm{softmax}\left(\frac{z}{T}\right) \,,

where :math:`z` is the vector of logits for each class produced
```

lorentzenchr commented on doc/modules/calibration.rst:
```diff
@@ -276,6 +276,54 @@ probabilities, the calibrated probabilities for each class
+where :math:`\left[z_1, \cdots, z_n \right]` are the logits for each class produced
+by the calibrating estimator for a given sample. In terms of scikit-learn's API,
+these correspond to the outputs of either :term:`decision_function` or
+:term:`predict_proba`. Probabilities are converted to logits by first adding a small
+positive constant to avoid taking the logarithm of zero, and then applying the natural
+logarithm.
```
```suggestion
:term:`predict_proba`. Probabilities are converted to logits by first adding a tiny
positive constant to avoid numerical issues with logarithm of zero, and then applying the natural
logarithm.
```

lorentzenchr commented on doc/modules/calibration.rst:
```diff
@@ -276,6 +276,54 @@ probabilities, the calibrated probabilities for each class
+positive constant to avoid taking the logarithm of zero, and then applying the natural
+logarithm.
+
+Temperature scaling is equivalent to training a neural network with a single neuron
+activated by the softmax function. The parameter :math:`T` is learned by minimizing
```
```suggestion
Temperature scaling is equivalent to training a neural network with a single neuron
activated by the softmax function and the classifier output as input to the net. The parameter :math:`T` is learned by minimizing
```
For me, we could also remove this neural net part.

lorentzenchr commented on doc/modules/calibration.rst:
```diff
@@ -276,6 +276,54 @@ probabilities, the calibrated probabilities for each class
+1. It provides a natural way to obtain multi-class probabilities, avoiding the
+   common dilemma of choosing between "One-vs-One" and "One-vs-Rest" schemes.
+
+2. It offers a computationally efficient way to generate probability predictions.
+
+   For example, :class:`~sklearn.svm.SVC` requires the parameter `probability=True`
+   in order to use :term:`predict_proba`, which increases training time.
+   Similarly, :class:`~sklearn.linear_model.SGDClassifier` supports :term:`predict_proba`
+   only when the `loss` parameter is set to `log_loss` or `modified_huber`.
+
+   Temperature scaling, in contrast, requires training only a single scalar parameter
+   to produce probability estimates.
```
```suggestion
```
In my understanding, this is simply not correct.

lorentzenchr commented on doc/modules/calibration.rst:
```diff
@@ -276,6 +276,54 @@ probabilities, the calibrated probabilities for each class
+scaling does not alter the accuracy of the calibrating estimator.
+
+
+Advantages of temperature scaling include:
+
+1. It provides a natural way to obtain multi-class probabilities, avoiding the
+   common dilemma of choosing between "One-vs-One" and "One-vs-Rest" schemes.
```
```suggestion
The big advantages of temperature scaling over other calibration methods is that it
provides a natural way to obtain (better) calibrated multi-class probabilities with just one free parameter in contrast to using a "One-vs-Rest" scheme that adds more parameters for each single class.
```

virchan committed:
update `calibration.rst` by applying reviewer's feedback.

virchan committed:
update `_TemperatureScaling.predict` method by applying reviewer's feedback.

virchan committed:
update `test_temperature_scaling` by applying code-review suggestions.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
update `test_temperature_scaling`.

virchan reviewed (commented):
The CI fails when checking that the ROC AUCs are equal up to 7 decimal places. I'll fix it later.

virchan committed:
removed error-raisings in private functions

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan committed:
updated ROC-AUC checking in the `test_temperature_Scaling` function

virchan reviewed (commented):
CI passed!

lorentzenchr commented on sklearn/tests/test_calibration.py:
```diff
@@ -359,7 +377,7 @@ def test_calibration_prefit(csr_container):
 
 
-@pytest.mark.parametrize("method", ["sigmoid", "isotonic"])
+@pytest.mark.parametrize("method", ["sigmoid", "isotonic", "temperature"])
```
```suggestion
@pytest.mark.parametrize(
    ["method", "calibrator"],
    [
        ("sigmoid", _SigmoidCalibration()),
        ("isotonic", IsotonicRegression(out_of_bounds="clip")),
        ("temperature", _TemperatureScaling())
    ]
)
```
and then remove the if-else inside.

lorentzenchr commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +426,90 @@ def test_sigmoid_calibration():
         _SigmoidCalibration().fit(np.vstack((exF, exF)), exY)
 
 
+@pytest.mark.parametrize(
+    "clf",
+    [LogisticRegression(penalty=None, tol=1e-6, random_state=0)],
+)
```
As this is only one estimator, we can remove the parametrize and put it inside the test function.

lorentzenchr commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +426,90 @@ def test_sigmoid_calibration():
+        assert_allclose(
+            roc_auc_score(y_cal, y_scores, multi_class="ovr"),
+            roc_auc_score(y_cal, y_scores_cal, multi_class="ovr"),
+        )
```
For logistic regression, we could even check that temperature is 1 if fitted on the training set.

lorentzenchr reviewed (commented):
Close to the finish line.

virchan committed:
Update `calibration.rst` by applying review suggestions.

virchan commented on sklearn/calibration.py:
```diff
@@ -683,41 +736,25 @@ def _fit_calibrator(clf, predictions, y, classes, method, sample_weight=None):
 
 
 class _CalibratedClassifier:
-    """Pipeline-like chaining a fitted classifier and its fitted calibrators.
-
-    Parameters
-    ----------
-    estimator : estimator instance
-        Fitted classifier.
-
-    calibrators : list of fitted estimator instances
-        List of fitted calibrators (either 'IsotonicRegression' or
-        '_SigmoidCalibration'). The number of calibrators equals the number of
-        classes. However, if there are 2 classes, the list contains only one
-        fitted calibrator.
-
-    classes : array-like of shape (n_classes,)
-        All the prediction classes.
-
-    method : {'sigmoid', 'isotonic'}, default='sigmoid'
-        The method to use for calibration. Can be 'sigmoid' which
-        corresponds to Platt's method or 'isotonic' which is a
-        non-parametric approach based on isotonic regression.
-    """
-
```
Oh, I see. I had thought [#31068 (comment)](https://github.com/scikit-learn/scikit-learn/pull/31068#discussion_r2086241540) applied to this case as well. I'll go ahead and revert it.

virchan committed:
Update `calibration.py` docstrings by applying review suggestions.

virchan committed:
Update `_convert_to_logits` by tightening `eps`.

virchan committed:
Update `test_temperature_scaling` to check that the temperature parameter is close to 1.0 when fitted on the training set of LogisticRegression.

virchan committed:
Merge branch 'main' into issues/28574_implementing_temperature_scaling

virchan committed:
Update `calibration.py`

virchan commented on sklearn/tests/test_calibration.py:
```diff
@@ -401,6 +426,91 @@ def test_sigmoid_calibration():
+            roc_auc_score(y_cal, y_scores_cal, multi_class="ovr"),
+        )
+
+        y_scores_train = clf.predict_proba(X_train)
+        ts = _TemperatureScaling().fit(y_scores_train, y_train)
+        assert_allclose(ts.beta_, 1.0, atol=1.1e-7, rtol=0)
```
> For logistic regression, we could even check that temperature is 1 if fitted on the training set.

Let me know if atol or rtol needs to be adjusted here.

virchan committed:
Update `_temperature_scaling` docstrings.

virchan reviewed (commented):
I updated the user guide and the docstrings in `calibration.py`. I also modified the test to check that the temperature parameter is close to 1 when the temperature scaler is fitted on the training set of the `LogisticRegression` classifier.

There are still some comments that need to be addressed, and I'll work on them later.

virchan committed:
Update `test_calibration.py`

virchan committed:
Merge branch 'main' into issues/28574_implementing_temperature_scaling

virchan committed:
Move `_temperature_scaling` into `_TemperatureScaling().fit` method.

virchan committed:
Move `_temperature_scaling` into `_TemperatureScaling().fit` method.

virchan committed:
Fix docstrings in `calibration.py`

virchan committed:
Update the `_fit_calibrator` function

virchan committed:
Place `response_method_name` logic inside `_convert_to_logits`.

virchan committed:
Place `response_method_name` logic inside `_fit_calibrator` and `_CalibratedClassifier.predict_proba` only.

virchan committed:
Merge remote-tracking branch 'origin/issues/28574_implementing_temperature_scaling' into issues/28574_implementing_temperature_scaling

# Conflicts:
#	sklearn/calibration.py

virchan committed:
Place `response_method_name` logic in `_fit_calibrator` and `_CalibratedClassifier.predict_proba` only.

virchan committed:
Remove blank line.

virchan committed:
Update `test_calibration.py`

virchan reviewed (commented):
I've refactored the part for checking `reponse_method_name`:
```python
if len(classes) == 2 and predictions.shape[-1] == 1:
    response_method_name = _check_response_method(
        clf,
        ["decision_function", "predict_proba"],
    ).__name__
    if response_method_name == "predict_proba":
        predictions = np.hstack([1 - predictions, predictions])
```

I think this only needs to be applied in two places: : `_fit_calibrator` and `_CalibratedClassifier.predict_proba`. But please let me know if there's a better way to handle this.

I've also moved `_temperature_scaling` inside `_TemperatureScaling.fit`.

CI has passed, so it's ready for review!

lorentzenchr commented on sklearn/calibration.py:
```diff
@@ -756,27 +796,37 @@ def predict_proba(self, X):
+                    proba, denominator, out=uniform_proba, where=denominator != 0
+                )
+        elif self.method == "temperature":
+            if n_classes == 2 and predictions.shape[-1] == 1:
+                response_method_name = _check_response_method(
+                    self.estimator,
+                    ["decision_function", "predict_proba"],
+                ).__name__
+                if response_method_name == "predict_proba":
+                    predictions = np.hstack([1 - predictions, predictions])
```
Why not put it into `_TemperatureScaling.predict`?

virchan committed:
Remove the `_temperature_scaling` function and add its logic into the `_TemperatureScaling.fit` method directly.

virchan committed:
Add `response_method_name` parameter to `_TemperatureScaling`.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan reviewed (commented):
I updated `_convert_to_logits` to handle the conversion to a 2D array in `_TemperatureScaling.fit` and `_TemperatureScaling.predict` when the input is 1D.

`_TemperatureScaling` now has a new parameter, `response_method_name` (`"decision_function"` or `"predict_proba"`), which indicates whether it should fit or predict based on the output of `decision_function` or `predict_proba`, respectively. The default value is `"decision_function"`.

In the `CalibratedClassifierCV` workflow, this value is computed in `_fit_calibrator`, then passed to `_TemperatureScaling` when the calibrator is initialised. If the input is 1D, `_convert_to_logits` will interpret it as either probabilities or decision values, accordingly.

This is to address an edge case where the output of `predict_proba` is 1D and was incorrectly converted to logits using `[-x, x]`, which was first caught by the `test_calibration_prefit` function.

Previously, I attempted something similar in [<code>2769eab</code>](https://github.com/scikit-learn/scikit-learn/pull/31068/commits/2769eab3c165b541e9569d1329fd96e4a665deea), but I thought it was awkward for `_TemperatureScaling` to store `response_method_name`, so I didn't finalise it at the time. We'll see how this version goes.

Edit: Reverted.

virchan committed:
Revert "Add `response_method_name` parameter to `_TemperatureScaling`."

This reverts commit c4ec0e8993363740463a253997d08779c8913885.

virchan committed:
Merge remote-tracking branch 'upstream/main' into issues/28574_implementing_temperature_scaling

virchan reviewed (commented):
The `response_method_name` commit has been reverted, and CI has passed.

lorentzenchr committed:
Update sklearn/calibration.py

lorentzenchr committed:
Update sklearn/calibration.py

lorentzenchr committed:
Update sklearn/calibration.py

virchan committed:
Merge branch 'main' into issues/28574_implementing_temperature_scaling

virchan commented:
Thank you everyone for your time! 🫡
